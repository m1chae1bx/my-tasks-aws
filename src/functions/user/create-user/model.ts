import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface CreateUserRequest {
  body: {
    id: string;
    email: string;
    password: string;
    fullName: string;
    nickname: string;
  };
}

export const createUserRequestSchema: JSONSchemaType<CreateUserRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    body: {
      type: "object",
      properties: {
        id: {
          type: "string",
          maxLength: 100,
        },
        email: {
          type: "string",
        },
        password: {
          type: "string",
        },
        fullName: {
          type: "string",
        },
        nickname: {
          type: "string",
          maxLength: 100,
        },
      },
      required: ["id", "email", "password", "fullName", "nickname"],
    },
  },
  required: ["body"],
};

export const validateRequest = ajv.compile(createUserRequestSchema);
