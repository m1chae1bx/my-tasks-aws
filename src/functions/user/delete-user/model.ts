import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface DeleteUserRequest {
  pathParameters: {
    userId: string;
  };
  body: {
    password: string;
  };
}

export const deleteUserRequestSchema: JSONSchemaType<DeleteUserRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
          maxLength: 100,
        },
      },
      required: ["userId"],
    },
    body: {
      type: "object",
      properties: {
        password: {
          type: "string",
        },
      },
      required: ["password"],
    },
  },
  required: ["pathParameters", "body"],
};

export const validateRequest = ajv.compile(deleteUserRequestSchema);
