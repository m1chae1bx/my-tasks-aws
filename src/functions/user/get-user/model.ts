import { JSONSchemaType, ajv } from "@libs/generic/util";

export interface GetUserRequest {
  pathParameters: {
    userId: string;
  };
}

export const getUserRequestSchema: JSONSchemaType<GetUserRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
          maxLength: 100,
        },
      },
      required: ["userId"],
    },
  },
  required: ["pathParameters"],
};

export const validateRequest = ajv.compile(getUserRequestSchema);
