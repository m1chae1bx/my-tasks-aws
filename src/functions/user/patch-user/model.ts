import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface PatchUserRequest {
  pathParameters: {
    userId: string;
  };
  body: {
    email?: string;
    fullName?: string;
    nickname?: string;
  };
}

export const patchUserRequestSchema: JSONSchemaType<PatchUserRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        userId: {
          type: "string",
          maxLength: 100,
        },
      },
      required: ["userId"],
    },
    body: {
      type: "object",
      properties: {
        email: {
          type: "string",
          nullable: true,
        },
        fullName: {
          type: "string",
          nullable: true,
        },
        nickname: {
          type: "string",
          maxLength: 100,
          nullable: true,
        },
      },
      additionalProperties: false,
    },
  },
  required: ["pathParameters", "body"],
};

export const validateRequest = ajv.compile(patchUserRequestSchema);
