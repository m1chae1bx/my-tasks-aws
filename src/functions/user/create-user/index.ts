import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { User } from "/opt/nodejs/user.model";
import { genericErrorHandler, isAWSError } from "/opt/nodejs/util";

const emailUnavailableResponse = {
  message: "Email address is already registered",
  code: "emailUnavailable",
};

const usernameUnavailableResponse = {
  message: "Username is already taken",
  code: "usernameUnavailable",
};

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      body: event.body ? JSON.parse(event.body) : null,
    };

    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad request",
          errors: validateRequest.errors,
        }),
      };
    }

    const { id, email, password, fullName, nickname } = request.body;

    const userByEmail = await User.getByEmail(email);
    if (userByEmail) {
      return {
        statusCode: 409,
        body: JSON.stringify(emailUnavailableResponse),
      };
    } else {
      const userByUsername = await User.get(id);
      if (userByUsername) {
        return {
          // TODO create a Response object or use existing AWS if any?
          statusCode: 409,
          body: JSON.stringify(usernameUnavailableResponse),
        };
      } else {
        const newUser = new User(id, email, fullName, nickname);
        newUser.setPassword(password);
        await newUser.save();

        const response = {
          message: `User ${id} was created successfully`,
          id: id,
        };
        return {
          statusCode: 200,
          body: JSON.stringify(response),
        };
      }
    }
  } catch (error) {
    if (isAWSError(error) && error.code === "ConditionalCheckFailedException") {
      return {
        statusCode: 409,
        body: JSON.stringify(emailUnavailableResponse),
      };
    }

    return genericErrorHandler(
      error,
      `An error occurred while creating the user. Please try again later.`
    );
  }
};
