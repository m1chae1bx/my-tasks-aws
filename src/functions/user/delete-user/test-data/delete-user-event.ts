import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";

export const deleteUserEvent = {
  ...baseAPIGatewayProxyEvent,
  pathParameters: {
    userId: "testUserId",
  },
  body: JSON.stringify({
    password: "secret-password",
  }),
};
