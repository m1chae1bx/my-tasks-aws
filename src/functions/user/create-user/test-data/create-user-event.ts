import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";
import { UserDetails } from "/opt/nodejs/user.model";

export const createUserEvent = {
  ...baseAPIGatewayProxyEvent,
  body: JSON.stringify({
    fullName: "Full Name",
    nickname: "Nickname",
    id: "testUserId",
    email: "test@email.com",
    password: "pass1234%",
  }),
};

export const validationErrors = [
  {
    instancePath: "/body/password",
    schemaPath: "#/properties/body/properties/password/minLength",
    keyword: "minLength",
    params: {
      limit: 8,
    },
    message: "must NOT have fewer than 8 characters",
  },
  {
    instancePath: "/body/password",
    schemaPath: "#/properties/body/properties/password/pattern",
    keyword: "pattern",
    params: {
      pattern:
        "^(?=[^A-Za-z]*[A-Za-z])(?=[^0-9]*[0-9])(?=[^@#$%&!_:\\-.]*[@#$%&!_:\\-.]).*",
    },
    message:
      'must match pattern "^(?=[^A-Za-z]*[A-Za-z])(?=[^0-9]*[0-9])(?=[^@#$%&!_:\\-.]*[@#$%&!_:\\-.]).*"',
  },
];

export const testUser: UserDetails = {
  id: "testUserId",
  email: "test@email.com",
  fullName: "Full Name",
  nickname: "Nickname",
  salt: "salt",
  hash: "hash",
};
