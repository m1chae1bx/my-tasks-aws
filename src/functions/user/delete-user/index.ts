import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { User } from "/opt/nodejs/user.model";
import { genericErrorHandler } from "/opt/nodejs/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      pathParameters: event.pathParameters,
      body: event.body ? JSON.parse(event.body) : null,
    };

    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad request",
          errors: validateRequest.errors,
        }),
      };
    }

    const userId = request.pathParameters.userId;
    const password = request.body.password;

    const user = await User.get(userId);
    if (!user) {
      return {
        statusCode: 404,
        body: JSON.stringify({
          message: "User not found",
          code: "notFound",
        }),
      };
    }
    if (!user.validatePassword(password)) {
      return {
        statusCode: 401,
        body: JSON.stringify({
          message: "Incorrect password",
          code: "incorrectPassword",
        }),
      };
    }

    await user.delete();

    return {
      statusCode: 200,
      body: JSON.stringify({
        message: `User ${userId} was deleted successfully`,
        code: "userDeleted",
      }),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      "An error occurred while deleting the user. Please try again later.",
      "deleteUserError"
    );
  }
};
