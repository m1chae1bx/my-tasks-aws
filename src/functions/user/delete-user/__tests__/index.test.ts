import { APIGatewayProxyResult } from "aws-lambda";
import { handler } from "../index";
import { validateRequest } from "../model";
import { deleteUserEvent } from "../test-data/delete-user-event";
import { User } from "/opt/nodejs/user.model";
import { ajvError } from "@test-utils/base-objects";
import { genericErrorHandler } from "/opt/nodejs/util";

jest.mock("../model", () => ({
  validateRequest: jest.fn(),
}));

jest.mock("/opt/nodejs/user.model", () => ({
  User: {
    get: jest.fn(),
  },
}));

jest.mock("/opt/nodejs/util", () => ({
  genericErrorHandler: jest.fn(),
}));

describe("delete-user", () => {
  describe("happy path", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (User.get as jest.Mock).mockResolvedValueOnce({
        validatePassword: jest.fn().mockReturnValueOnce(true),
        delete: jest.fn(),
      });
      result = await handler(deleteUserEvent);
    });

    it("should return a 200 status code", () => {
      expect(result.statusCode).toBe(200);
    });

    it("should return a success message", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  describe("sad path - invalid request", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(false);
      validateRequest.errors = ajvError;
      result = await handler(deleteUserEvent);
    });

    it("should return a 400 status code", () => {
      expect(result.statusCode).toBe(400);
    });

    it("should return errors", () => {
      expect(result.body).toMatchSnapshot();
      expect(result.body).toContain(ajvError[0].message);
    });
  });

  describe("sad path - invalid request (no body)", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(false);
      validateRequest.errors = ajvError;
      result = await handler({ ...deleteUserEvent, body: null });
    });

    it("should return a 400 status code", () => {
      expect(result.statusCode).toBe(400);
    });

    it("should return errors", () => {
      expect(result.body).toMatchSnapshot();
      expect(result.body).toContain(ajvError[0].message);
    });
  });

  describe("sad path - user not found", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (User.get as jest.Mock).mockResolvedValueOnce(null);
      result = await handler(deleteUserEvent);
    });

    it("should return a 404 status code", () => {
      expect(result.statusCode).toBe(404);
    });

    it("should return an error message", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  describe("sad path - incorrect password", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (User.get as jest.Mock).mockResolvedValueOnce({
        validatePassword: jest.fn().mockReturnValueOnce(false),
      });
      result = await handler(deleteUserEvent);
    });

    it("should return a 401 status code", () => {
      expect(result.statusCode).toBe(401);
    });

    it("should return an error message", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  describe("sad path - error while deleting user", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (User.get as jest.Mock).mockResolvedValueOnce({
        validatePassword: jest.fn().mockReturnValueOnce(true),
        delete: jest.fn().mockRejectedValueOnce(new Error("error")),
      });
      (genericErrorHandler as jest.Mock).mockReturnValueOnce({
        statusCode: 500,
        body: "test error",
      });
      result = await handler(deleteUserEvent);
    });

    it("should return a 500 status code", () => {
      expect(result.statusCode).toBe(500);
    });

    it("should return an error message", () => {
      expect(result.body).toMatchSnapshot();
    });
  });
});
