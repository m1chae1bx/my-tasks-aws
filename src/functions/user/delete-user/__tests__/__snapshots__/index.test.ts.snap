// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delete-user happy path should return a success message 1`] = `"{\\"message\\":\\"User testUserId was deleted successfully\\",\\"code\\":\\"userDeleted\\"}"`;

exports[`delete-user sad path - error while deleting user should return an error message 1`] = `"test error"`;

exports[`delete-user sad path - incorrect password should return an error message 1`] = `"{\\"message\\":\\"Incorrect password\\",\\"code\\":\\"incorrectPassword\\"}"`;

exports[`delete-user sad path - invalid request (no body) should return errors 1`] = `"{\\"message\\":\\"Bad request\\",\\"errors\\":[{\\"keyword\\":\\"\\",\\"instancePath\\":\\"\\",\\"schemaPath\\":\\"\\",\\"params\\":{},\\"message\\":\\"test error\\"}]}"`;

exports[`delete-user sad path - invalid request should return errors 1`] = `"{\\"message\\":\\"Bad request\\",\\"errors\\":[{\\"keyword\\":\\"\\",\\"instancePath\\":\\"\\",\\"schemaPath\\":\\"\\",\\"params\\":{},\\"message\\":\\"test error\\"}]}"`;

exports[`delete-user sad path - user not found should return an error message 1`] = `"{\\"message\\":\\"User not found\\",\\"code\\":\\"notFound\\"}"`;
